osgp/shared/shared/src/main/java/org/opensmartgridplatform/shared/security/providers/AbstractEncryptionProvider.java

```diff
@@ -0,0 +1,98 @@
+import org.opensmartgridplatform.shared.security.EncryptedSecret;
+import org.opensmartgridplatform.shared.security.EncryptionProviderType;
+import org.opensmartgridplatform.shared.security.Secret;
+
+public abstract class AbstractEncryptionProvider {
+
+    private static final int BLOCK_SIZE=16;
+
+    protected File keyFile;
+
+    public abstract EncryptionProviderType getType();
+
+    protected abstract Cipher getCipher()
+            throws NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException;
+
+    protected abstract AlgorithmParameterSpec getAlgorithmParameterSpec();
+
+    protected abstract Key getSecretEncryptionKey(String keyReference, int cipherMode);
+
+    public void setKeyFile(File keyFile) {
+        this.keyFile = keyFile;
+    }
+
+    public EncryptedSecret encrypt(Secret secret, String keyReference) {
+        try {
+            final Cipher cipher = this.getCipher();
+            cipher.init(Cipher.ENCRYPT_MODE, this.getSecretEncryptionKey(keyReference, Cipher.ENCRYPT_MODE),
+                    this.getAlgorithmParameterSpec());
+            return new EncryptedSecret(this.getType(), cipher.doFinal(secret.getSecret()));
+        } catch (Exception e) {
+            //InvalidKeyException | IllegalBlockSizeException | BadPaddingException |
+            // InvalidAlgorithmParameterException |
+            //NoSuchPaddingException | NoSuchAlgorithmException | NoSuchProviderException
+            throw new EncrypterException("Could not encrypt secret with keyReference " + keyReference, e);
+        }
+    }
+
+    public Secret decrypt(EncryptedSecret secret, String keyReference) {
+
+        if (secret.getType() != this.getType()) {
+            throw new EncrypterException(
+                    "EncryptionProvider for type " + this.getType().name() + " cannot decrypt secrets of type "
+                            + secret.getType().name());
+        }
+
+        try {
+            final Cipher cipher = this.getCipher();
+            cipher.init(Cipher.DECRYPT_MODE, this.getSecretEncryptionKey(keyReference, Cipher.DECRYPT_MODE),
+                    this.getAlgorithmParameterSpec());
+            final byte[] decryptedData = cipher.doFinal(secret.getSecret());
+
+            if (this.checkNullBytesPrepended(decryptedData)) {
+                return new Secret(Arrays.copyOfRange(decryptedData, BLOCK_SIZE, decryptedData.length));
+            } else {
+                return new Secret(decryptedData);
+            }
+        } catch (Exception e) {
+            //InvalidKeyException | IllegalBlockSizeException | BadPaddingException |
+            // InvalidAlgorithmParameterException |
+            //NoSuchPaddingException | NoSuchAlgorithmException | NoSuchProviderException
+            throw new EncrypterException("Could not decrypt secret with keyReference " + keyReference, e);
+        }
+    }
+
+    private boolean checkNullBytesPrepended(final byte[] bytes) {
```

Comment:
return 
    bytes.length > BLOCK_SIZE 
    && 
    Arrays.stream(Arrays.copyOfRange(bytes, 0, BLOCK_SIZE)).allMatch(b -> b == 0);

Something like this?